File: 프로그래머스/lv3/43105. 정수 삼각형/정수 삼각형.cpp
#include <bits/stdc++.h>

#define MAX 501
using namespace std;

int dp[MAX][MAX];
int solution(vector<vector<int>> triangle) {
    int answer = 0;
    int height = triangle.size();
    dp[0][0] = triangle[0][0];
    
    for(int i = 1 ; i < triangle.size(); i++) {
        for(int j = 0 ; j < triangle[i].size() ; j++) {
            dp[i][j] = triangle[i][j];
            if(j-1 < 0) {
                dp[i][j] += dp[i-1][j];
            } else if(j >= triangle[i-1].size()) {
                dp[i][j] += dp[i-1][j-1];
            } else {
                dp[i][j] += max(dp[i-1][j-1], dp[i-1][j]);
            }
            
        }
    }
    answer = *max_element(dp[height-1], dp[height-1]+height);

    return answer;
}

File: 프로그래머스/3/43105. 정수 삼각형/정수 삼각형.js
function solution(triangle) {
    var answer = 0;
    let dp = [];
    
    triangle.forEach((cur, idx) => {
        if(idx === 0) {
            dp.push(cur);
            return;
        }
        
        const curDp = cur.reduce((acc, el, i) => {
            if (i === 0) {
                acc.push(dp[idx - 1][i] + el);
            } else if (i === cur.length - 1) {
                acc.push(dp[idx - 1][i - 1] + el);
            } else {
                acc.push(Math.max(dp[idx - 1][i - 1], dp[idx - 1][i]) + el);
            }
            return acc;
        }, []);
        
        dp.push(curDp);
    })
    
    answer = Math.max(...dp[dp.length - 1]);
    
    return answer;
}